clients/src/main/java/org/apache/kafka/common/record/Record.java

```diff
@@ -132,4 +121,16 @@
      * @return the array of headers
      */
     Header[] headers();
+
+    static void ensureValid(Record record) {
+        if (record instanceof AbstractLegacyRecordBatch) {
+            ((AbstractLegacyRecordBatch) record).ensureValid();
+        }
+    }
+
+    static boolean isValid(Record record) {
+        if (record instanceof AbstractLegacyRecordBatch) {
+            return ((AbstractLegacyRecordBatch) record).isValid();
+        } else return true;
```

Comment:
Why is this change an improvement?

Reply:
The method “isVaild” is meaningless for v2 and later. It seems to me it is not worth being a interface. If this change is overkill, alternative is that we can add default implementation to those stale methods. The v2 and later can skip implementation of stale methods.

Reply:
Polymorphism is a fine way to deal with that though. It's more confusing if we have a combination of polymorphism and pattern matching (well, the if/else version of it which doesn't include exhaustiveness checking).

I don't see an issue with having an implementation that always return true or is a no-op for v2. A default may be ok, but it introduces more coupling between the interface and implementations. It seems that `isValid` is only used by `DumpLogSegments` (outside of tests), so we can maybe remove that altogether. I would probably leave `ensureValid` as it was before this PR.

Reply:
thanks for nice suggestions. will copy that.